lib/rouge/lexers/make.rb

```diff
@@ -80,22 +107,22 @@ def initialize(opts={})
 
       state :shell do
         # macro interpolation
-        rule %r/\$\(\s*[a-z_]\w*\s*\)/i, Name::Variable
-        # $(shell ...)
-        rule %r/(\$\()(\s*)(shell)(\s+)/m do
+        rule %r/\$[({][\t ]*[a-z_][\w:=%.]*[\t ]*[)}]/i, Name::Variable
+        # function invocation
+        rule %r/(\$[({])([\t ]*)(#{Make.functions.join('|')})([\t ]+)/m do
```

Comment:
This regex looks a little broken. As written, it would allow whitespace between the `{` or `(` and the function name but not allow whitespace after the function name and the closing `}`. I say `}` and not `)` because it doesn't look like you can close this with a parenthesis. Finally, the tokens used aren't consistent. The opening `${` or `$(` is tokenised as `Name::Function` but the `)` is part of the `Name::Builtin` token.

Reply:
It appears that GNU make doesn't actually permit whitespace before the function name, but I left it in because the previous lexer permitted it, so perhaps other make tools permit it?

How I thought I understood this was that in a string such as `$( call user_function,arg )`, this rule consumes characters as follows:

Characters | Token
----------- | ------
`$(` | Name::Function
` ` | Text
`call` | Name::BuiltIn
` ` | Text

then `user_function,arg ` is matched next as the longest string until the `stop` lookahead matches, and is lexed independently. After that, the `)` matches the second rule in state `:shell_expr` and is given a Name::Function token.

Have I got this wrong?

It's maybe a bit of a cheat to lex any built-in function other than `shell` as though it was a shell script line, but it does have the advantage that it already handles variable expansions within function call invocations, which are a fairly common occurrence.

Reply:
Ugh, I'm a dummy. I misread that closing brace in the interpolation. OK, the regex is fine—mistake on my end.